File: solution/209MinimumSizeSubarraySum/209MinimumSizeSubarraySum.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
  int minSubArrayLen(int s, vector<int> &nums) {
    int n = nums.size();
    int min_pos = 0;
    int min_len = n;
    int i = 0;
    int j = 0;
    int sum = 0;
    while (j < n) {
      sum += nums[j];
      while (i < j && sum - nums[i] >= s) {
        sum -= nums[i];
        i++;
      }
      if (sum >= s && j - i + 1 < min_len) {
        min_len = j - i + 1;
        min_pos = i;
      }
      j++;
    }
    if (sum >= s) {
      return min_len;
    } else {
      return 0;
    }
  }
};

int main() {
  Solution s;
  vector<int> nums = {2, 3, 1, 2, 4, 3};
  cout << s.minSubArrayLen(7, nums) << endl;
  cout << s.minSubArrayLen(4, nums) << endl;
  return 0;
}

File: src/209MinimumSizeSubarraySum/209MinimumSizeSubarraySum.rs
impl Solution {
    pub fn min_sub_array_len(s: i32, nums: Vec<i32>) -> i32 {
        let (mut i, mut j, mut k) = (0, 0, nums.len()+1);
        
        let mut sum = 0;
        while j < nums.len() {      
            sum += nums[j];
            while sum >= s {
                k = std::cmp::min(k, j-i+1);
                sum -= nums[i];
                i += 1;
            }
            j+=1;
        }
        if k==nums.len()+1{
            0
        }else{
            k as i32
        }
    }
}
